### ISSUE TITLE
Feature: Add support for golang HashDir/h1/sha256 in fixed output derivations.

### ISSUE THREAD

**Issue by bergkvist** [NixOS/nix#8021]
**Is your feature request related to a problem? Please describe.**
Currently, a pure nix wrapping of golang modules is not possible (requiring `vendorSha256` or similar) - even when a golang lockfile (`go.sum`) is available. This is due to golang hashing directories in a different way compared to nix. (https://pkg.go.dev/golang.org/x/mod/sumdb/dirhash)

**Describe the solution you'd like**
If fixed output derivations added support for the golang hash-algorithm, this could massively improve the ergonomics of packaging go modules that have `go.sum` lockfiles.

```nix
mkDerivation {
  # ...
  outputHashAlgo = "golang:h1";
  outputHash = "h1:0CfEIISq7TuYL3j771MWULgwwjU+GofnZX9QAmXWZgo=";
}
```

**Comment by zowoq** [comment]
`go.sum` is not a lockfile.

**Comment by bergkvist** [comment]
@zowoq You are right, it is only a checksum file, not a lockfile. Regardless - I believe, supporting the directory hashing method used in the checksum file would be very useful for making `buildGoModule` more comfortable to use.

**Comment by Ericson2314** [comment]
I am actually OK with supporting a great many types of content-adddressing, but yes I agree that we should still only stick to well-specified ones, and if it is the case that Go's has issues, then yes that ties our hands :(.

**Comment by bergkvist** [comment]
From the source code, there is some more documentation:
```go
// Hash1 is the "h1:" directory hash function, using SHA-256.
//
// Hash1 is "h1:" followed by the base64-encoded SHA-256 hash of a summary
// prepared as if by the Unix command:
//
//	sha256sum $(find . -type f | sort) | sha256sum
//
// More precisely, the hashed summary contains a single line for each file in the list,
// ordered by sort.Strings applied to the file names, where each line consists of
// the hexadecimal SHA-256 hash of the file content,
// two spaces (U+0020), the file name, and a newline (U+000A).
//
// File names with newlines (U+000A) are disallowed.
```
https://cs.opensource.google/go/x/mod/+/refs/tags/v0.9.0:sumdb/dirhash/hash.go;l=44

**Comment by lheckemann** [comment]
I started implementing this a while back, but never finished it: https://github.com/NixOS/nix/compare/master...lheckemann:nix:go
